File: src/main/java/at/scheinecker/intellij/coco/reference/CocoProductionReference.java
package at.scheinecker.intellij.coco.reference;

import at.scheinecker.intellij.coco.CocoUtil;
import at.scheinecker.intellij.coco.psi.CocoFormalAttributes;
import at.scheinecker.intellij.coco.psi.HasCocoProductionReference;
import com.intellij.codeInsight.lookup.LookupElementBuilder;
import com.intellij.openapi.util.TextRange;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiReference;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public class CocoProductionReference extends AbstractRenamableReference<HasCocoProductionReference> implements PsiReference {

    public CocoProductionReference(@NotNull HasCocoProductionReference element, TextRange textRange) {
        super(element, textRange);
    }

    @Nullable
    @Override
    public PsiElement resolve() {
        return CocoUtil.findProduction(myElement.getContainingFile(), myElement.getName());
    }

    @NotNull
    @Override
    public Object[] getVariants() {
        return CocoUtil.findProductions(myElement.getContainingFile())
                .stream()
                .map(it -> {
                    CocoFormalAttributes formalAttributes = it.getFormalAttributes();
                    LookupElementBuilder lookupElementBuilder = LookupElementBuilder.create(it)
                            .withTypeText("Production");

                    if (formalAttributes != null) {
                        return lookupElementBuilder.withTailText(formalAttributes.getText(), true);
                    }

                    return lookupElementBuilder;
                })
                .toArray();
    }
}

File: src/main/kotlin/me/salzinger/intellij/coco/reference/CocoProductionReference.kt
package me.salzinger.intellij.coco.reference

import com.intellij.codeInsight.lookup.LookupElementBuilder
import com.intellij.openapi.util.TextRange
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiReference
import me.salzinger.intellij.coco.CocoUtil
import me.salzinger.intellij.coco.psi.HasCocoProductionReference

class CocoProductionReference(element: HasCocoProductionReference, textRange: TextRange) :
    AbstractRenamableReference<HasCocoProductionReference>(element, textRange), PsiReference {

    override fun resolve(): PsiElement? {
        return CocoUtil.findProduction(myElement.containingFile, myElement.name)
    }

    override fun getVariants(): Array<Any> {
        return CocoUtil.findProductions(myElement.containingFile)
            .map { it ->
                val formalAttributes = it.formalAttributes
                val lookupElementBuilder = LookupElementBuilder.create(it)
                    .withTypeText("Production")

                if (formalAttributes != null) {
                    return@map lookupElementBuilder.withTailText(formalAttributes.text, true)
                }

                lookupElementBuilder
            }
            .toTypedArray()
    }
}
